lib/nostrum/api.ex

```diff
@@ -778,6 +831,149 @@ defmodule Nostrum.Api do
     |> bangify
   end
 
+  @doc ~S"""
+  Gets a list of emojis for a given guild.
+
+  If the request is successful, this function returns `{:ok, emojis}`, where 
+  `emojis` is a list of `Nostrum.Struct.Emoji`. Otherwise, returns `{:error, reason}`.
+  """
+  @spec list_guild_emojis(Guild.id) :: error | {:ok, [Emoji.t]}
+  def list_guild_emojis(guild_id) do
+    request(:get, Constants.guild_emojis(guild_id))
+    |> handle_request_with_decode({:list, {:struct, Emoji}})
+  end
+
+  @doc ~S"""
+  Same as `list_guild_emojis/1`, but raises `Nostrum.Error.ApiError` in case of failure.
+  """
+  @spec list_guild_emojis!(Guild.id) :: no_return | [Emoji.t]
+  def list_guild_emojis!(guild_id) do
+    list_guild_emojis(guild_id)
+    |> bangify
+  end
+
+  @doc ~S"""
+  Gets an emoji for the given guild and emoji ids.
+
+  If the request is successful, this function returns `{:ok, emoji}`, where 
+  `emoji` is a `Nostrum.Struct.Emoji`. Otherwise, returns `{:error, reason}`.
+  """
+  @spec get_guild_emoji(Guild.id, Emoji.id) :: error | {:ok, Emoji.t}
+  def get_guild_emoji(guild_id, emoji_id) do
+    request(:get, Constants.guild_emoji(guild_id, emoji_id))
+    |> handle_request_with_decode({:struct, Emoji})
+  end
+
+  @doc ~S"""
+  Same as `get_guild_emoji/2`, but raises `Nostrum.Error.ApiError` in case of failure.
+  """
+  @spec get_guild_emoji!(Guild.id, Emoji.id) :: no_return | Emoji.t
+  def get_guild_emoji!(guild_id, emoji_id) do
+    get_guild_emoji(guild_id, emoji_id)
+    |> bangify
+  end
+
+  @doc ~S"""
+  Creates a new emoji for the given guild.
+
+  ## Options 
+ 
+    * `:name` (string) - name of the emoji
+    * `:image` (base64 data URI) - the 128x128 emoji image. Maximum size of 256kb
+    * `:roles` (list of `t:Nostrum.Struct.Snowflake.t/0`) - roles for which this emoji will be whitelisted
+    (default: [])
+
+  `:name` and `:image` are always required.
+
+  ## Return Values
+
+  If the request is successful, this function returns `{:ok, emoji}`, where 
+  `emoji` is a `Nostrum.Struct.Emoji`. Otherwise, returns `{:error, reason}`.
+
+  ## Examples
+
+  ```Elixir
+  image = "data:image/png;base64,YXl5IGJieSB1IGx1a2luIDQgc3VtIGZ1az8="
+
+  Nostrum.Api.create_guild_emoji(43189401384091, name: "nostrum", image: image, roles: [])
+  ```
+  """
+  @spec create_guild_emoji(Guild.id, keyword | map) :: error | {:ok, Emoji.t}
```

Comment:
Hmm, throughout this module we'll have `keyward | map` in specs whenever there are optional parameters. Do you think we should similarly include something like an `optional` type that describes the map/keyword list functionality we have here?

If we decide to do something like this feel free to open an issue so we can keep track of it. We probably don't want to go back to the `user` functions in this pr and change them.

Reply:
I'll create an issue specifically for tracking these. However, I'll be changing the `keyword | map` types for the current Emoji-related endpoints.